vochain/ist/ist.go

```diff
@@ -0,0 +1,298 @@
+func (c *Controller) Schedule(height uint32, id []byte, action Action) error {
+	if id == nil {
+		return fmt.Errorf("cannot schedule IST action: nil id")
+	}
+	actions, err := c.Actions(height)
+	if err != nil {
+		return err
+	}
+	// set the IST action
+	actions[string(id)] = action
+
+	// store the IST actions
+	log.Debugw("schedule IST action", "height", height, "id", fmt.Sprintf("%x", id), "action", ActionsToString[action.ID])
+	return c.storeToNoState(dbIndex(height), actions.encode())
+}
+
+// Actions returns the IST actions scheduled for the given height.
+// If no actions are scheduled, it returns an empty IstActions.
+func (c *Controller) Actions(height uint32) (Actions, error) {
+	// get the IST actions
+	actions := Actions{}
+	actionsBytes, err := c.retrieveFromNoState(dbIndex(height))
+	if err != nil {
+		if errors.Is(err, db.ErrKeyNotFound) {
+			// no IST actions scheduled for this height, we return empty
+			actions = make(map[string]Action)
+			return actions, nil
+		}
+		return nil, fmt.Errorf("cannot get IST actions on height %d: %w", height, err)
+	}
+	// decode the IST actions
+	if err := actions.decode(actionsBytes); err != nil {
+		return nil, fmt.Errorf("could not decode actions: %w", err)
+	}
+	return actions, nil
+}
+
+// Reschedule reschedules an IST action to the given new height.
+func (c *Controller) Reschedule(id []byte, oldHeight, newHeight uint32) error {
+	if len(id) == 0 {
+		return fmt.Errorf("cannot reschedule IST action: nil id")
+	}
+	if oldHeight >= newHeight {
+		return fmt.Errorf("cannot reschedule IST action: old height (%d) >= new height (%d)", oldHeight, newHeight)
+	}
+	actions, err := c.Actions(oldHeight)
+	if err != nil {
+		return err
+	}
+	// find the IST action
+	_, ok := actions[string(id)]
+	if !ok {
+		return fmt.Errorf("cannot reschedule IST action: action not found")
+	}
+	// store the IST action
+	a := actions[string(id)]
+	a.Attempts++
+	if err := c.Schedule(newHeight, id, a); err != nil {
+		return err
+	}
+	// delete the IST action
+	delete(actions, string(id))
+
+	// update the IST actions for the old height
+	if err := c.storeToNoState(dbIndex(oldHeight), actions.encode()); err != nil {
+		return fmt.Errorf("cannot reschedule IST action: %w", err)
+	}
+	return nil
+}
+
+// Commit executes the IST actions scheduled for the given height.
+func (c *Controller) Commit(height uint32, isSynchronizing bool) error {
+	actions, err := c.Actions(height)
+	if err != nil {
+		return fmt.Errorf("cannot commit IST actions: %w", err)
+	}
+	if len(actions) == 0 {
+		return nil
+	}
+	for id, action := range actions {
+		switch action.ID {
+		case ActionComputeResults:
+			log.Debugw("compute results", "height", height, "id", fmt.Sprintf("%x", id), "attempt", action.Attempts)
+			// check if we have all the revealed keys, else reschedule
+			if !c.checkRevealedKeys(action.ElectionID) {
+				// we are missing some keys, we cannot compute the results yet
+				newHeight := height + (1 + action.Attempts*2)
```

Comment:
Maybe it is useful creating a function for this. Something like `RescheduleHeight()`